platform/linux-generic/odp_traffic_mngr.c

```diff
@@ -2246,12 +2246,14 @@ static void tm_send_pkt(tm_system_t *tm_system, uint32_t max_sends)
 			tm_egress_marking(tm_system, odp_pkt);
 
 		tm_system->egress_pkt_desc = EMPTY_PKT_DESC;
-		if (tm_system->egress.egress_kind == ODP_TM_EGRESS_PKT_IO)
-			odp_pktout_send(tm_system->pktout, &odp_pkt, 1);
-		else if (tm_system->egress.egress_kind == ODP_TM_EGRESS_FN)
+		if (tm_system->egress.egress_kind == ODP_TM_EGRESS_PKT_IO) {
+			if (odp_pktout_send(tm_system->pktout, &odp_pkt, 1) != 1)
+				odp_packet_free(odp_pkt);
```

Comment:
This looks ok as it fixes the packet leak and thus makes the code more correct than it was.

That said, this tm_send_pkt() looks a bit fishy to me as it does not seem to be designed to handle the case when the pktout cannot absorb all the packets sent through it. I would think TM should then slow down sending and let the configured schedulers and queue management policies do their thing and not simply drop packets here (or within the odp_pktout_send). But that does not seem to be the case and it seems to me that TM works only when a shaper is limiting the output rate and not the pktout itself.

Maybe the TM code here assumed that odp_pktout_send() will not normally fail and will block when there congestion in the output link. But I am not sure if TM would work correctly even if odp_pktout_send() worked that way.

So, I am ok with this fix as it fixes the packet leak and does not break anything, although I am not sure if the action taken in the failure case should be something else (retry?) than freeing the packet if we want to fix other things. Assuming there are other things to fix and I am not just confused. The potential other issues can be fixed separately.